File: pull/puller.go
package pull

import (
	"log"
	"net/url"
)

// Remote is the interface for pull sources.
type Remote interface {
	PullArtifacts(env string, name string, localWarehouse string, opt Options) (int, error)
}

func newRemote(s string) Remote {
	u, err := url.Parse(s)
	if err != nil {
		log.Panicf("pull: parsing remote url error: %v", err)
	}

	switch u.Scheme {
	case "", "s3":
		return NewS3Remote(s)
	default:
		log.Panicf("pull: unknown remote scheme: %s", u.Scheme)
	}

	return nil
}

File: src/pull/puller.js
import { S3Remote } from "./remote-s3.js";

export function newRemote(value) {
  const url = new URL(value.includes("://") ? value : `s3://${value}`);

  switch (url.protocol.replace(/:$/, "")) {
    case "":
    case "s3":
      return new S3Remote(value);
    default:
      throw new Error(`pull: unknown remote scheme: ${url.protocol.replace(/:$/, "")}`);
  }
}
